lib/body.js

```diff
@@ -19,215 +13,190 @@ module.exports = Body;
-	}
-
-	// no charset in content type, peek at response body for at most 1024 bytes
-	if (!res && this._raw.length > 0) {
-		for (var i = 0; i < this._raw.length; i++) {
-			str += this._raw[i].toString()
-			if (str.length > 1024) {
-				break;
-			}
-		}
-		str = str.substr(0, 1024);
-	}
-
-	// html5
-	if (!res && str) {
-		res = /<meta.+?charset=(['"])(.+?)\1/i.exec(str);
-	}
-
-	// html4
-	if (!res && str) {
-		res = /<meta[\s]+?http-equiv=(['"])content-type\1[\s]+?content=(['"])(.+?)\2/i.exec(str);
-
-		if (res) {
-			res = /charset=(.*)/i.exec(res.pop());
-		}
-	}
-
-	// xml
-	if (!res && str) {
-		res = /<\?xml.+?encoding=(['"])(.+?)\1/i.exec(str);
-	}
-
-	// found charset
-	if (res) {
-		charset = res.pop();
-
-		// prevent decode issues when sites use incorrect encoding
-		// ref: https://hsivonen.fi/encoding-menu/
-		if (charset === 'gb2312' || charset === 'gbk') {
-			charset = 'gb18030';
-		}
-	}
-
-	// turn raw buffers into utf-8 string
-	return convert(
-		Buffer.concat(this._raw)
-		, encoding
-		, charset
-	).toString();
-
-};
-
-/**
- * Clone body given Res/Req instance
- *
- * @param   Mixed  instance  Response or Request instance
- * @return  Mixed
- */
-Body.prototype._clone = function(instance) {
-	var p1, p2;
-	var body = instance.body;
-
-	// don't allow cloning a used body
-	if (instance.bodyUsed) {
-		throw new Error('cannot clone body after it is used');
-	}
-
-	// check that body is a stream and not form-data object
-	// note: we can't clone the form-data object without having it as a dependency
-	if (bodyStream(body) && typeof body.getBoundary !== 'function') {
-		// tee instance body
-		p1 = new PassThrough();
-		p2 = new PassThrough();
-		body.pipe(p1);
-		body.pipe(p2);
-		// set instance body to teed body and return the other teed body
-		instance.body = p1;
-		body = p2;
-	}
-
-	return body;
+class Body {
+    constructor(body, opts) {
+        opts = opts || {};
+        this._rawBody = body;
+        this.bodyUsed = false;
+        this.size = opts.size || 0;
+        this.timeout = opts.timeout || 0;
+        this.url = opts.url;
+    }
+
+    get body() {
+        const rawBody = this._rawBody;
+        if (!rawBody) {
+            return null;
+        } else if (rawBody instanceof webStreams.ReadableStream) {
+            return rawBody;
+        } else if (typeof rawBody === 'string'
+                || Buffer.isBuffer(rawBody)
+                || isNodeStream(rawBody) && rawBody.readable && !rawBody.getBoundary) {
+            // Convert to ReadableStream.
+            this._rawBody = webStreams.toWebReadableStream(rawBody);
+            return this._rawBody;
+        } else if (rawBody && rawBody.getBoundary) {
+            // FormData instance. Make an exception, for now.
+            return rawBody;
+        } else {
+            throw new TypeError(`Unsupported Response body type: ${typeof body}`);
+        }
+    }
+
+    set body(val) {
+        this._rawBody = val;
+    }
```

Comment:
This goes against the WHATWG spec, which specifies that

```webidl
readonly attribute ReadableStream? body;
```

The `readonly` annotation means that the setter should be left blank.